Move crc32_fold_copy helper functions into test harness.

nmoinvaz opened the pull request:
These are not needed for the zlib implementation as it defaults to `crc32_fold_copy_c` which calls the `crc32` for the current architecture and then memcpy.

nmoinvaz commented on test/test_crc32_fold_copy.cc:
```diff
@@ -9,6 +9,7 @@ extern "C" {
 #  include "zbuild.h"
 #  include "arch_functions.h"
 #  include "test_cpu_features.h"
+#  include "crc32_fold_copy.h"
```
fixed. included "arch_functions.h".

nmoinvaz commented on test/crc32_fold_copy.h:
```diff
@@ -0,0 +1,33 @@
+#include "zbuild.h"
```
Nah.

nmoinvaz commented on test/crc32_fold_copy.h:
```diff
@@ -0,0 +1,33 @@
+        memcpy(dst, src, len);
+    }
+#  endif
+#endif
```
fixed.

nmoinvaz committed:
Add missing crc32_fold_copy functions in gtest_zlib.

phprus commented:
Removing specialized implementations of crc32_fold_copy results in an extra indirect call:

https://github.com/zlib-ng/zlib-ng/blob/6e6ec1c33cb2b183b31dbe3ccf85cba5e6db1f85/arch/generic/crc32_fold_c.c#L16

Won't this lead to a performance penalty?

nmoinvaz committed:
Remove crc32_fold C functions

nmoinvaz committed:
Move crc32_fold_copy helper functions into test harness.

These are not needed for the zlib implementation as it defaults to
crc32_fold_copy_c which calls the crc32 for the current architecture
and then memcpy.

nmoinvaz committed:
Check functable crc32_fold functions against NULL to fix warnings

nmoinvaz committed:
Rename crc32_fold to crc32_fold_update.

nmoinvaz commented on arch/x86/x86_functions.h:
```diff
@@ -76,14 +76,14 @@ uint32_t adler32_fold_copy_avx512_vnni(uint32_t adler, uint8_t *dst, const uint8
 uint32_t crc32_fold_vpclmulqdq_reset(crc32_fold *crc);
 void     crc32_fold_vpclmulqdq_copy(crc32_fold *crc, uint8_t *dst, const uint8_t *src, size_t len);
-void     crc32_fold_vpclmulqdq(crc32_fold *crc, const uint8_t *src, size_t len, uint32_t init_crc);
+void     crc32_fold_vpclmulqdq_update(crc32_fold *crc, const uint8_t *src, size_t len, uint32_t init_crc);
```
Fixed.

nmoinvaz committed:
Remove crc32_fold reset and final C functions in test harness.

phprus commented on deflate.c:
```diff
@@ -1030,7 +1042,9 @@ int32_t Z_EXPORT PREFIX(deflate)(PREFIX3(stream) *strm, int32_t flush) {
 #ifdef GZIP
     if (s->wrap == 2) {
-        strm->adler = FUNCTABLE_CALL(crc32_fold_final)(&s->crc_fold);
+        if (FUNCTABLE_CALL(crc32_fold_final) != NULL) {
+            strm->adler = FUNCTABLE_CALL(crc32_fold_final)(&s->crc_fold);
+        }
```
As far as I can see, `crc32_fold` is only used on x86.
Perhaps it would make sense to add an `#ifdef` and disable runtime checks for `crc32_fold* != NULL` and the field in the `internal_state` structure on non-x86 platforms?

nmoinvaz replied:
I think I am going to try a different route. First remove templates from x86 crc32 code. Second convert `crc32_fold` into `crc32_copy` and see if the performance hit is worth it.
